### PR TITLE
feat: add observedGeneration attribute to the status conditions

### PR DESCRIPTION
**Pull request by tremes**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._
This is adding a new `ObservedGeneration int64 `json:"observedGeneration,omitempty"` to the `PrometheusCondition`. 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [X] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

**Comment by tremes** [comment]
I am experimenting with extending the [testPromDegradedConditionStatus](https://github.com/prometheus-operator/prometheus-operator/blob/main/test/e2e/prometheus_test.go#L4407) to compare the `observedGeneration` value before and after marking it as degraded.

**Comment by simonpasquier** [comment]
lgtm on the implementation.

> I am experimenting with extending the [testPromDegradedConditionStatus](https://github.com/prometheus-operator/prometheus-operator/blob/main/test/e2e/prometheus_test.go#L4407) to compare the observedGeneration value before and after marking it as degraded.

IMHO we need also (and more importantly) to check in [WaitForPrometheusReady](https://github.com/prometheus-operator/prometheus-operator/blob/ad64869358bfc470bac56c48a1ecee6dfb8dc88c/test/framework/prometheus.go#L379) that that .spec.generation == .status.condtions[].observedGeneration.

File: test/framework/prometheus.go [hunk 1]
```diff
@@ -427,12 +431,12 @@ func (f *Framework) WaitForPrometheusReady(ctx context.Context, p *monitoringv1.
 			return false, nil
 		}
 
-		if reconciled.Status != monitoringv1.PrometheusConditionTrue {
+		if available.Status != monitoringv1.PrometheusConditionTrue {
```
Comment (simonpasquier) [inline test/framework/prometheus.go:434]: oops

File: test/framework/prometheus.go [hunk 2]
```diff
@@ -405,6 +405,10 @@ func (f *Framework) WaitForPrometheusReady(ctx context.Context, p *monitoringv1.
 			if cond.Type == monitoringv1.PrometheusReconciled {
 				reconciled = &cond
 			}
+			if cond.ObservedGeneration != current.Generation {
+				pollErr = errors.Errorf("prometheus generation is not equal to conditions observed generation")
```
Comment (simonpasquier) [inline test/framework/prometheus.go:409]: let's report the exact problem.
```suggestion
				pollErr = errors.Errorf("observed generation %d for condition %s isn't equal to the state generation %d", cond.ObservedGeneration, cond.Type, current.Generaiton)
```

File: pkg/apis/monitoring/v1/types.go [hunk 3]
```diff
@@ -630,6 +630,10 @@ type PrometheusCondition struct {
 	// Human-readable message indicating details for the condition's last transition.
 	// +optional
 	Message string `json:"message,omitempty"`
+	// observedGeneration represents the .metadata.generation that the condition was set based upon.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:633]:
```suggestion
	// ObservedGeneration represents the .metadata.generation that the condition was set based upon.
```

**Review by simonpasquier — COMMENTED** [review]
Only a few nits.

**Pull request merged**